parsons/databases/database_connector.py

```diff
@@ -0,0 +1,108 @@
+from abc import ABC, abstractmethod
+from typing import Optional
+from parsons.etl.table import Table
+
+
+class DatabaseConnector(ABC):
+    """
+    An abstract base class that provides a uniform interface for all Parsons database connectors.
+    This class should be used in functions instead of the specific database connector classes
+    when the functions don't rely on database-specific functionality.
+
+    It ensures that any class that inherits from it implements the `table_exists`, `copy`, and
+    `query` methods, which are common operations when working with databases.
+
+    Note:
+        The Python type system (as of 3.10.6) will not stop you from breaking the type contract
+        of method signatures when implementing a subclass. It is up to the author of a database
+        connector to ensure that it satisfies this interface. Be careful to, for example, not
+        change the types of the parameters or leave out optional parameters that are specified
+        in the interface.
```

Comment:
I'm gonna take a gander and read some more about options here. Would love to see if there's a package or some other way to enforce things. I would also love to learn more about the extra optional parameters. That seems like something we'll want to support and then have descriptive errors for.

Reply:
Please let me know if you found anything. As far as I could tell, the ecosystem of Python type-checking tools isn't to this level of sophistication at static-time analysis.

The ABC module does raise a runtime error if you instantiate classes that inherit from an ABC and haven't defined all of the methods. But from what I was able to determine, it doesn't check the type signature of those methods. I'm not surprised, because the Python runtime isn't aware of the static time type annotations at all.

I agree this would be great to have.

Reply:
Yeah we've talked a bit about trying to add more type hints/type checking to Parsons. I'd love to see that too but it seems like a separate issue that should not block this PR.